[#227] Migrate from tfsec to Trivy

nvminhtue opened the pull request:
- Close #227 

## What happened 👀

Integrate [Trivy](https://aquasecurity.github.io/trivy) that will replace the current `tfsec`

## Insight 📝

- Integrate the `trivy` config and replace all `tfsec`'s ignorance.
- Add generating project step and run `trivy` scanner on.
- Remove Ros from reviewer

## Proof Of Work 📹
Generate the completed AWS services locally and detect no HIGH or CRITICAL issues.
![image](https://github.com/nimblehq/infrastructure-templates/assets/42540264/b57b7e65-9884-4420-a0ad-638bf6beec63)


The example of a failure check without putting the trivy ignores
![image](https://github.com/nimblehq/infrastructure-templates/assets/42540264/25818f36-ad63-4b0e-9bc0-ad5403342af6)

CI will be failed if the Trivy scan found any HIGH or CRITICAL severity
![image](https://github.com/nimblehq/infrastructure-templates/assets/42540264/15255a89-5bce-49e0-8976-4eb5f6e50133)

hoangmirs commented on trivy.yaml:
Should we also have another file for the generated project? So we can test trivy on that project 🤔

nvminhtue replied:
Nice one, updated alongside with the document in [60e3b42](https://github.com/nimblehq/infrastructure-templates/pull/262/commits/60e3b42aa7e3fd0b9a6a99e56fa61db243914ef2)

Nihisil replied:
@nvminhtue why we need trivy configuration for template project itself? 🤔 it is not clear part for me

trivy can check typescript code as well?

nvminhtue replied:
no, it is just a config file to base the check on the unbundled terraform file from the root.
I agree that it causes confusion, removed it in [b22f6f4](https://github.com/nimblehq/infrastructure-templates/pull/262/commits/b22f6f410338fc7dc12445e142ab928cf2345919)

hoangmirs reviewed (commented):
Could you add some guidelines to run trivy on local?

Nihisil commented on trivy.yaml:
```diff
@@ -0,0 +1,28 @@
+    - shared/.terraform/
+
+  scanners:
+    - vuln
```
does it have `vuln` scanner for terraform?

from this page https://aquasecurity.github.io/trivy/v0.47/docs/coverage/iac/terraform/

they have only misconfiguration and secret scanners

nvminhtue replied:
I think they are missing somehow because they're having the section for that
https://aquasecurity.github.io/trivy/v0.47/docs/scanner/vulnerability/

The config file setup can be found here: https://aquasecurity.github.io/trivy/v0.47/docs/references/configuration/config-file/

Nihisil commented:
@nvminhtue it looks like something is not working properly.

1. I have generated new project from your branch
2. Deleted few `trivy:ignore` lines there and there

When I'm running trivy locally I'm receiving exit error without any details:
```
$ trivy config .
2023-12-30T11:40:37.723+0700	INFO	Loaded trivy.yaml
2023-12-30T11:40:37.736+0700	INFO	Misconfiguration scanning is enabled
2023-12-30T11:40:38.522+0700	INFO	Detected config files: 13
exit 1
```

When trivy run on CI there are no errors at all, CI run is green:
```
Running Trivy with trivy.yaml config from:  trivy.yaml
2023-12-30T04:37:57.267Z	INFO	Loaded trivy.yaml
2023-12-30T04:37:57.276Z	INFO	"--dependency-tree" only shows the dependents of vulnerable packages. Note that it is the reverse of the usual dependency tree, which shows the packages that depend on the vulnerable package. It supports limited package managers. Please see the document for the detail.
2023-12-30T04:37:57.276Z	WARN	"--dependency-tree" can be used only with "--format table".
[20](https://github.com/Nihisil/test-infra/actions/runs/7362083108/job/20040144448#step:6:21)23-12-30T04:37:57.283Z	INFO	Need to update DB
2023-12-30T04:37:57.283Z	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
2023-12-30T04:37:57.283Z	INFO	Downloading DB...
32.83 MiB / 42.08 MiB [----------------------------------------------->_____________] 78.02% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 15.43 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 15.43 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 15.43 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 14.44 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 14.44 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 14.44 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [-------------------------------------------------] 100.00% 24.40 MiB p/s 1.9s2023-12-30T04:37:59.724Z	INFO	Vulnerability scanning is enabled
20[23](https://github.com/Nihisil/test-infra/actions/runs/7362083108/job/20040144448#step:6:24)-12-30T04:37:59.724Z	INFO	Secret scanning is enabled
2023-12-30T04:37:59.7[24](https://github.com/Nihisil/test-infra/actions/runs/7362083108/job/20040144448#step:6:25)Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-12-30T04:37:59.724Z	INFO	Please see also https://aquasecurity.github.io/trivy/v0.45/docs/scanner/secret/#recommendation for faster secret detection
2023-12-30T04:37:59.740Z	INFO	Number of language-specific files: 0
```

I expected to see errors locally and in CI for the blocks where I removed the ignore statements. Could you please help me figure out what's wrong with my tests?

nvminhtue commented:
> @nvminhtue it looks like something is not working properly.
> 
> 1. I have generated new project from your branch
> 2. Deleted few `trivy:ignore` lines there and there
> 
> When I'm running trivy locally I'm receiving exit error without any details:
> 
> ```
> $ trivy config .
> 2023-12-30T11:40:37.723+0700	INFO	Loaded trivy.yaml
> 2023-12-30T11:40:37.736+0700	INFO	Misconfiguration scanning is enabled
> 2023-12-30T11:40:38.522+0700	INFO	Detected config files: 13
> exit 1
> ```
> 
> When trivy run on CI there are no errors at all, CI run is green:
> 
> ```
> Running Trivy with trivy.yaml config from:  trivy.yaml
> 2023-12-30T04:37:57.267Z	INFO	Loaded trivy.yaml
> 2023-12-30T04:37:57.276Z	INFO	"--dependency-tree" only shows the dependents of vulnerable packages. Note that it is the reverse of the usual dependency tree, which shows the packages that depend on the vulnerable package. It supports limited package managers. Please see the document for the detail.
> 2023-12-30T04:37:57.276Z	WARN	"--dependency-tree" can be used only with "--format table".
> [20](https://github.com/Nihisil/test-infra/actions/runs/7362083108/job/20040144448#step:6:21)23-12-30T04:37:57.283Z	INFO	Need to update DB
> 2023-12-30T04:37:57.283Z	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
> 2023-12-30T04:37:57.283Z	INFO	Downloading DB...
> 32.83 MiB / 42.08 MiB [----------------------------------------------->_____________] 78.02% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 15.43 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 15.43 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 15.43 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 14.44 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 14.44 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 14.44 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [-------------------------------------------------] 100.00% 24.40 MiB p/s 1.9s2023-12-30T04:37:59.724Z	INFO	Vulnerability scanning is enabled
> 20[23](https://github.com/Nihisil/test-infra/actions/runs/7362083108/job/20040144448#step:6:24)-12-30T04:37:59.724Z	INFO	Secret scanning is enabled
> 2023-12-30T04:37:59.7[24](https://github.com/Nihisil/test-infra/actions/runs/7362083108/job/20040144448#step:6:25)Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
> 2023-12-30T04:37:59.724Z	INFO	Please see also https://aquasecurity.github.io/trivy/v0.45/docs/scanner/secret/#recommendation for faster secret detection
> 2023-12-30T04:37:59.740Z	INFO	Number of language-specific files: 0
> ```
> 
> I expected to see errors locally and in CI for the blocks where I removed the ignore statements. Could you please help me figure out what's wrong with my tests?

@Nihisil, 
It actually generates the errors into the `trivy-output.json` file.
But it's a good catch, the CI never notices if there is any mis-configurated, updated the trivy config file in [d5a52da](https://github.com/nimblehq/infrastructure-templates/pull/262/commits/d5a52da6e9140937b5b161a3ec20377cca149e91) and [f6e6628](https://github.com/nimblehq/infrastructure-templates/pull/262/commits/f6e6628083b5de02b4394bb6f72637e41614f276)

Nihisil commented:
@nvminhtue thank you. After your fix I can receive error on localhost, but CI is still the green.

Also I have deleted all trivy ignore lines, and it shows me only one error. Do we need to keep ignore lines in that case, or it is misconfiguration and it should show more errors?

And another issue is that it doesn't show where the error is, based on error message it is impossible to find out where the fix is required. Is it possible to adjust it to show file and line number?

local output:
```$ trivy config .
2024-01-03T10:56:32.454+0700	INFO	Loaded trivy.yaml
2024-01-03T10:56:32.462+0700	INFO	Misconfiguration scanning is enabled
2024-01-03T10:56:35.821+0700	INFO	Detected config files: 26

 (terraform)

Tests: 1 (SUCCESSES: 0, FAILURES: 1, EXCEPTIONS: 0)
Failures: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

MEDIUM: Cluster does not have Deletion Protection enabled
═══════════════════════════════════════════════════════════════════════════════════════════════════════════
Ensure deletion protection is enabled for RDS clusters.

See https://avd.aquasec.com/misconfig/n/a
───────────────────────────────────────────────────────────────────────────────────────────────────────────


exit 1
```

CI output:
```
Running Trivy with trivy.yaml config from:  trivy.yaml
2024-01-03T03:57:34.451Z	INFO	Loaded trivy.yaml
2024-01-03T03:57:34.460Z	WARN	"--list-all-pkgs" cannot be used with "--format table". Try "--format json" or other formats.
2024-01-03T03:57:34.460Z	INFO	"--dependency-tree" only shows the dependents of vulnerable packages. Note that it is the reverse of the usual dependency tree, which shows the packages that depend on the vulnerable package. It supports limited package managers. Please see the document for the detail.
[20](https://github.com/Nihisil/test-infra/actions/runs/7393288283/job/20112986384#step:6:21)24-01-03T03:57:34.465Z	INFO	Need to update DB
2024-01-03T03:57:34.465Z	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
2024-01-03T03:57:34.465Z	INFO	Downloading DB...
26.84 MiB / 42.08 MiB [-------------------------------------->______________________] 63.78% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 25.38 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 25.38 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 25.38 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 23.74 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 23.74 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 23.74 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [-------------------------------------------------] 100.00% 25.48 MiB p/s 1.9s2024-01-03T03:57:36.788Z	INFO	Vulnerability scanning is enabled
2024-01-03T03:57:36.788Z	INFO	Secret scanning is enabled
20[24](https://github.com/Nihisil/test-infra/actions/runs/7393288283/job/20112986384#step:6:25)-01-03T03:57:36.788Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2024-01-03T03:57:36.788Z	INFO	Please see also https://aquasecurity.github.io/trivy/v0.45/docs/scanner/secret/#recommendation for faster secret detection
2024-01-03T03:57:36.800Z	INFO	Number of language-specific files: 0
```

Nihisil commented on .github/wiki/Trivy-local-running.md:
```diff
@@ -0,0 +1,12 @@
+This project is using Trivy as a vulnerability scanner to replace the role of `tfsec` with some extra benefits:
+1. Access to more languages and features in the same tool.
+2. Access to more integrations with tools and services through the rich ecosystem around Trivy.
+3. Commercially supported by Aqua as well as by a the passionate Trivy community. tfsec will continue to remain available for the time being, although our engineering attention will be directed at Trivy going forward.
```
> our engineering attention

it is misleading sentence because we are not developing tfsec or trivy

nvminhtue replied:
It introduces the reason we migrated to `trivy` from `tfsec` and shows beneficial points. 
Removed the third point to keep things short in [b22f6f4](https://github.com/nimblehq/infrastructure-templates/pull/262/commits/b22f6f410338fc7dc12445e142ab928cf2345919)

nvminhtue commented:
> @nvminhtue thank you. After your fix I can receive error on localhost, but CI is still the green.
> 
> Also I have deleted all trivy ignore lines, and it shows me only one error. Do we need to keep ignore lines in that case, or it is misconfiguration and it should show more errors?
> 
> And another issue is that it doesn't show where the error is, based on error message it is impossible to find out where the fix is required. Is it possible to adjust it to show file and line number?
> 
> local output:
> 
> ```
> 2024-01-03T10:56:32.454+0700	INFO	Loaded trivy.yaml
> 2024-01-03T10:56:32.462+0700	INFO	Misconfiguration scanning is enabled
> 2024-01-03T10:56:35.821+0700	INFO	Detected config files: 26
> 
>  (terraform)
> 
> Tests: 1 (SUCCESSES: 0, FAILURES: 1, EXCEPTIONS: 0)
> Failures: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 1, HIGH: 0, CRITICAL: 0)
> 
> MEDIUM: Cluster does not have Deletion Protection enabled
> ═══════════════════════════════════════════════════════════════════════════════════════════════════════════
> Ensure deletion protection is enabled for RDS clusters.
> 
> See https://avd.aquasec.com/misconfig/n/a
> ───────────────────────────────────────────────────────────────────────────────────────────────────────────
> 
> 
> exit 1
> ```
> 
> CI output:
> 
> ```
> Running Trivy with trivy.yaml config from:  trivy.yaml
> 2024-01-03T03:57:34.451Z	INFO	Loaded trivy.yaml
> 2024-01-03T03:57:34.460Z	WARN	"--list-all-pkgs" cannot be used with "--format table". Try "--format json" or other formats.
> 2024-01-03T03:57:34.460Z	INFO	"--dependency-tree" only shows the dependents of vulnerable packages. Note that it is the reverse of the usual dependency tree, which shows the packages that depend on the vulnerable package. It supports limited package managers. Please see the document for the detail.
> [20](https://github.com/Nihisil/test-infra/actions/runs/7393288283/job/20112986384#step:6:21)24-01-03T03:57:34.465Z	INFO	Need to update DB
> 2024-01-03T03:57:34.465Z	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
> 2024-01-03T03:57:34.465Z	INFO	Downloading DB...
> 26.84 MiB / 42.08 MiB [-------------------------------------->______________________] 63.78% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [----------------------------------------------------------->] 100.00% ? p/s ?42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 25.38 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 25.38 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 25.38 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 23.74 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 23.74 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [---------------------------------------------->] 100.00% 23.74 MiB p/s ETA 0s42.08 MiB / 42.08 MiB [-------------------------------------------------] 100.00% 25.48 MiB p/s 1.9s2024-01-03T03:57:36.788Z	INFO	Vulnerability scanning is enabled
> 2024-01-03T03:57:36.788Z	INFO	Secret scanning is enabled
> 20[24](https://github.com/Nihisil/test-infra/actions/runs/7393288283/job/20112986384#step:6:25)-01-03T03:57:36.788Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
> 2024-01-03T03:57:36.788Z	INFO	Please see also https://aquasecurity.github.io/trivy/v0.45/docs/scanner/secret/#recommendation for faster secret detection
> 2024-01-03T03:57:36.800Z	INFO	Number of language-specific files: 0
> ```

Not sure how did you try on your local, everything works fine on my end
![image](https://github.com/nimblehq/infrastructure-templates/assets/42540264/85f7e8dc-8497-4049-8133-e283f73c0fa2)

To make sure we are on the same page, have you generated a local template and removed the `trivy:ignore` then?


The CI won't run since I couldn't find the way to unbundle the template by script, hence no tf files to run through, I believe we can do that on a separated story.

Nihisil commented:
> To make sure we are on the same page, have you generated a local template and removed the trivy:ignore then?

Yep, this is what I did.

Can you please check it on this branch: https://github.com/Nihisil/test-infra/tree/test-trivy

It shows an error that I provided above, and I'm not sure where the issue is.

Nihisil commented:
> The CI won't run since I couldn't find the way to unbundle the template by script, hence no tf files to run through, I believe we can do that on a separated story.

We need to ensure it runs on CI, or we should remove it from the linting step of GH workflow job and create a ticket to add it later. Otherwise, there might be confusion where developers might think that Trivy validation is already working in CI

nvminhtue commented:
> > The CI won't run since I couldn't find the way to unbundle the template by script, hence no tf files to run through, I believe we can do that on a separated story.
> 
> We need to ensure it runs on CI, or we should remove it from the linting step of GH workflow job and create a ticket to add it later. Otherwise, there might be confusion where developers might think that Trivy validation is already working in CI

I created the story [here](https://github.com/nimblehq/infrastructure-templates/issues/272), will work on that first and apply the change to this one, which can ensure that `trivy` is successfully integrated.

Nihisil commented:
I'm sorry that I wasn't clear earlier. Trivy isn't working with the generated project, neither on localhost nor on CI for me. This is most important part.

As for the ticket that you created, we have it already: https://github.com/nimblehq/infrastructure-templates/issues/181

nvminhtue commented:
> I'm sorry that I wasn't clear earlier. Trivy isn't working with the generated project, neither on localhost nor on CI for me. This is most important part.
> 
> As for the ticket that you created, we have it already: #181

Thanks for pointing me to that story 👍 
About your concern, I tried to pull your repo and it works on my end, seems there are some missing packages that I might not added to the CI.
![image](https://github.com/nimblehq/infrastructure-templates/assets/42540264/b638b895-f728-4a61-a3be-7d604fd57d7c)
Checking on that shortly

Nihisil commented:
maybe for local trivy installation we need to do some setup? I just did `asdf install` and after that tried to use it through `trivy config .`

nvminhtue commented:
> maybe for local trivy installation we need to do some setup? I just did `asdf install` and after that tried to use it through `trivy config .`

Yes, that's pretty much enough, it should work after then 🤔

Nihisil commented:
> Yes, that's pretty much enough, it should work after then 🤔

I've reinstalled Trivy, and it's working smoothly on localhost now. Thanks for your patience 🙏 

The only one remained concern for this PR is not working Trivy on CI for generated project

Nihisil commented on .github/workflows/test-generated-project.yml:
this file and other ones are from another PR, can you please clean up it

nvminhtue replied:
Since the containing PR will be merged in advance, we also need that to test the `Trivy` run as well, I opt to keep those files 👍

Nihisil commented:
@nvminhtue please rebase this PR with develop branch to solve merge conflicts :pray:

malparty commented on .github/wiki/Trivy-local-running.md:
````diff
@@ -0,0 +1,11 @@
+2. Access to more integrations with tools and services through the rich ecosystem around Trivy.
+
+## Trivy Local Scan
+```bash
````
````suggestion

```bash
````

Minor markdown lint suggestion

malparty commented on .github/wiki/_Sidebar.md:
```diff
@@ -15,3 +15,4 @@
 - [[Testing]]
 - [[Modify the Infrastructure Diagram | Modify infra diagram]]
 - [[Publishing]]
+- [[Trivy Local Running]]
```
```suggestion
- [[Trivy Local Run]]
```

or `Running Trivy Locally`? :)

malparty commented on .github/workflows/test-generated-project.yml:
```diff
@@ -39,17 +39,12 @@ jobs:
-        uses: hashicorp/setup-terraform@v2
-        with:
-          terraform_version: ${{ env.TERRAFORM_VERSION }}
+      - name: Install dependencies in .tool-versions
```
```suggestion
      - name: Install dependencies from .tool-versions
```

(we don't install the dependencies in the `.tool-versions` file, but we use the file to install the dependencies).

malparty commented on .github/workflows/test-generated-project.yml:
```diff
@@ -39,17 +39,12 @@ jobs:
-        with:
-          version: ${{ env.TFSEC_VERSION }}
-
+      - name: Run trivy linter
```
```suggestion
      - name: Run trivy scanner
```

According to the doc, Trivy is more referred to as a scanner, rather than a linter 💭

malparty commented on templates/addons/versionControl/github/.github/workflows/lint.yml:
```diff
@@ -25,16 +21,11 @@ jobs:
-        uses: hashicorp/setup-terraform@v2
-        with:
-          terraform_version: ${{ env.TERRAFORM_VERSION }}
+      - name: Install dependencies in .tool-versions
```
```suggestion
      - name: Install dependencies from .tool-versions
```

malparty commented on templates/addons/versionControl/github/.github/workflows/lint.yml:
```diff
@@ -25,16 +21,11 @@ jobs:
-        uses: aquasecurity/tfsec-action@v1.0.3
-        with:
-          version: ${{ env.TFSEC_VERSION }}
+      - name: Run trivy linter
```
```suggestion
      - name: Run trivy scanner
```

nvminhtue replied:
Update them all in [a518570](https://github.com/nimblehq/infrastructure-templates/pull/262/commits/a51857096f78c1142f953c3f54351c668daf93a0)

malparty reviewed (approved):
Approved ahead with minor suggestions :white_check_mark:
